File: apps/android/app/src/main/kotlin/com/maroonedsoftware/deadair/station/StationProbe.kt
package com.maroonedsoftware.deadair.station

import com.maroonedsoftware.deadair.sdk.DeadairSdk
import com.maroonedsoftware.deadair.sdk.runtime.SdkError
import kotlinx.serialization.MissingFieldException
import kotlinx.serialization.SerializationException
import java.security.cert.CertificateException
import javax.net.ssl.SSLHandshakeException

/** What asking an address whether it is a station got back. */
sealed interface StationCheck {
    /** It answered, and this is what it calls itself. */
    data class Reachable(val stationName: String) : StationCheck

    /**
     * A deadair station, answering a shape this app does not know.
     *
     * Told apart from "not a station" because the two are fixed differently and the wrong message
     * sends somebody to check an address that was right all along. `MissingFieldException` is the
     * signal: it means the body WAS a JSON object and simply lacked fields the contract requires,
     * which is what an older API looks like. A web page or a router's login form fails to parse as
     * JSON at all and lands in the case below.
     */
    data class Incompatible(val missing: String?) : StationCheck

    /** Something is there and it is not a deadair station: a 404, a login page, a router's UI. */
    data class NotAStation(val status: Int?) : StationCheck

    /** Nothing answered: wrong host, wrong port, no network, TLS refused. */
    data class Unreachable(val cause: String?) : StationCheck

    /**
     * Something answered, but this phone does not trust the certificate it presented.
     *
     * Told apart from [Unreachable] because the fix is different: a self-signed or private-CA
     * certificate is an ordinary thing for a self-hosted station to have, and the fix is
     * installing that CA on the phone rather than checking the address or the network.
     */
    data class Untrusted(val cause: String?) : StationCheck
}

/**
 * Whether an address is a station, asked before the app agrees to remember it.
 *
 * Through `/nowplaying` because it is the one deliberately public route: it needs no session, it
 * answers 200 on a quiet station rather than a 404, and it names the station — so one call both
 * proves the address and gives the listener something to recognise. A typo that reaches somebody
 * else's web server is told apart from a typo that reaches nothing, because those are different
 * mistakes and the listener fixes them differently.
 */
class StationProbe(private val sdkFor: (StationUrl) -> DeadairSdk) {
    suspend fun check(station: StationUrl): StationCheck =
        try {
            StationCheck.Reachable(sdkFor(station).nowplaying.getNowPlaying().station)
        } catch (error: Exception) {
            val untrustedCause = error.untrustedCertificateCause()
            when {
                untrustedCause != null ->
                    // The SDK or Ktor may wrap the TLS failure in something else entirely (even
                    // in SdkError, if a proxy in between answers first) so this is checked ahead
                    // of every other branch rather than caught by type.
                    StationCheck.Untrusted(untrustedCause.message)
                error is SdkError ->
                    // A status the contract does not describe. Something is listening on this
                    // address; it is just not a station.
                    StationCheck.NotAStation(error.status)
                error is MissingFieldException ->
                    // JSON, and an object, and missing something the contract requires. That is a
                    // station running an API this app does not match: almost always one that has
                    // not been redeployed since the client was built.
                    StationCheck.Incompatible(error.missingFields.firstOrNull())
                error is SerializationException ->
                    // A body that would not parse as the contract at all: a web page where JSON
                    // was expected, or JSON of some entirely different shape.
                    StationCheck.NotAStation(null)
                else -> StationCheck.Unreachable(error.message)
            }
        }
}

/**
 * Walks this throwable's cause chain looking for a TLS trust failure, because the SDK or Ktor may
 * wrap it in something else before it reaches [StationProbe.check]. A visited set guards against a
 * cause that cycles back on itself.
 */
private fun Throwable.untrustedCertificateCause(): Throwable? {
    val seen = mutableSetOf<Throwable>()
    var current: Throwable? = this
    while (current != null && seen.add(current)) {
        if (current is SSLHandshakeException || current is CertificateException) return current
        current = current.cause
    }
    return null
}

File: apps/ios/Packages/DeadairCore/Sources/DeadairCore/Station/StationProbe.swift
import DeadairSdk
import Foundation

/// What asking an address whether it is a station got back.
public enum StationCheck: Equatable, Sendable {
    /// It answered, and this is what it calls itself.
    case reachable(stationName: String)

    /// A deadair station, answering a shape this app does not know.
    ///
    /// Told apart from `notAStation` because the two are fixed differently, and the wrong message
    /// sends somebody to check an address that was right all along. A JSON object missing a key
    /// the contract requires is what an older API looks like; a web page or a router's login form
    /// fails to parse as the contract at all and lands in the case below.
    case incompatible(missing: String?)

    /// Something is there and it is not a deadair station: a 404, a login page, a router's UI.
    case notAStation(status: Int?)

    /// Nothing answered: wrong host, wrong port, no network.
    case unreachable(cause: String?)

    /// Something answered, but this phone does not trust the certificate it presented.
    ///
    /// Told apart from `unreachable` because the fix is different: a self-signed or private-CA
    /// certificate is an ordinary thing for a self-hosted station to have, and the fix is
    /// installing and trusting that CA on the phone rather than checking the address.
    case untrusted(cause: String?)
}

/// Whether an address is a station, asked before the app agrees to remember it.
///
/// Through `/nowplaying` because it is the one deliberately public route: it needs no session, it
/// answers 200 on a quiet station rather than a 404, and it names the station, so one call both
/// proves the address and gives the listener something to recognise.
public struct StationProbe: Sendable {
    private let sdkFor: @Sendable (StationUrl) -> Deadair

    public init(sdkFor: @escaping @Sendable (StationUrl) -> Deadair) {
        self.sdkFor = sdkFor
    }

    public func check(_ station: StationUrl) async -> StationCheck {
        do {
            return .reachable(stationName: try await sdkFor(station).nowplaying.getNowPlaying().station)
        } catch {
            return Self.classify(error)
        }
    }

    /// The outcome for a failure, checked in the order that makes each branch true.
    static func classify(_ error: Error) -> StationCheck {
        // First, ahead of every other branch: URLSession may wrap a TLS failure in another error,
        // and the fix it needs is not the one any other message suggests.
        if let untrusted = untrustedCertificate(in: error as NSError) {
            return .untrusted(cause: untrusted.localizedDescription)
        }
        if let error = error as? SdkError {
            // No response at all.
            if error.status == 0 { return .unreachable(cause: error.message) }
            // A 2xx here means the body did not decode: the generated client wraps the decoder's
            // error in an `SdkError` that keeps the body but not the reason, so the body is
            // decoded again to find out which kind of wrong it is.
            if (200..<300).contains(error.status) { return decodeFailure(in: error.body) }
            // A status the contract does not describe. Something is listening on this address; it
            // is just not a station.
            return .notAStation(status: error.status)
        }
        if let error = error as? DecodingError { return decodeFailure(error) }
        return .unreachable(cause: error.localizedDescription)
    }

    private static func decodeFailure(in body: Data) -> StationCheck {
        do {
            _ = try SdkJSON.makeDecoder().decode(NowPlaying.self, from: body)
            // It decodes now, so the failure was something else about the response.
            return .notAStation(status: nil)
        } catch let error as DecodingError {
            return decodeFailure(error)
        } catch {
            return .notAStation(status: nil)
        }
    }

    private static func decodeFailure(_ error: DecodingError) -> StationCheck {
        // JSON, and an object, and missing something the contract requires: a station running an
        // API this app does not match, almost always one not redeployed since the app was built.
        if case .keyNotFound(let key, _) = error { return .incompatible(missing: key.stringValue) }
        // A body that would not parse as the contract at all.
        return .notAStation(status: nil)
    }

    /// The certificate failures `URLSession` reports, found anywhere in the underlying-error chain.
    private static let certificateCodes: Set<Int> = [
        NSURLErrorServerCertificateHasBadDate,
        NSURLErrorServerCertificateUntrusted,
        NSURLErrorServerCertificateHasUnknownRoot,
        NSURLErrorServerCertificateNotYetValid,
        NSURLErrorClientCertificateRejected,
        NSURLErrorClientCertificateRequired,
    ]

    private static func untrustedCertificate(in error: NSError) -> NSError? {
        var current: NSError? = error
        while let candidate = current {
            if candidate.domain == NSURLErrorDomain, certificateCodes.contains(candidate.code) { return candidate }
            current = candidate.userInfo[NSUnderlyingErrorKey] as? NSError
        }
        return nil
    }
}
